File: Level 2/Rabbit Hole 1/java/Rabbit_Hole_1.java
import java.util.*;

class Solution {
  
  public int getMaxVisitableWebpages(int N, int[] L) {
    int maxVisits = 0;
    
    Map<Integer, Integer> webpageHops = new HashMap<>();
    LinkedList<Integer> visitOrder = new LinkedList<>();
    
    for (int i = 0; i < N; i++) {
      if (webpageHops.containsKey(i)) {
        continue;
      }
      
      Set<Integer> visited = new HashSet<>();
      visited.add(i);
      visitOrder.add(i);
      
      int nextPos = L[i] - 1;
      int size = 1;
      while (!visited.contains(nextPos)) {
        if (webpageHops.containsKey(nextPos)) {
          size += webpageHops.get(nextPos);
          break;
        }
        
        size++;
        visited.add(nextPos);
        visitOrder.add(nextPos);
        nextPos = L[nextPos] - 1;
      }
      
      if (size > maxVisits) {
        maxVisits = size;
      }
      
      boolean reachedCycle = false;
      while (visitOrder.size() > 0) {
        Integer page = visitOrder.poll();
        
        if (page == nextPos) {
          reachedCycle = true;
        }
        webpageHops.put(page, size);
        
        if (!reachedCycle) {
          size--;
        }
      }
      
      if (webpageHops.size() == N || maxVisits == N) {
        break;
      }
    }
    
    return maxVisits;
  }
  
}

File: Level 2/Rabbit Hole 1/python/Rabbit_Hole_1.py
from typing import List
from collections import deque

def getMaxVisitableWebpages(N: int, L: List[int]) -> int:
    maxVisitableWebpages = 0
        
    webpageHops = {}  # key: webpage index, value: number of webpages that can be visited from this webpage
    visitOrder = deque()  # order of webpages visited from each starting page (populated and cleared as we go)
    for i in range(N):
        if i in webpageHops:
            # We've already visited this webpage and calculated the number of webpages that can be visited from it
            continue

        # Start from webpage i and visit all webpages that can be visited from it
        visited = {i}
        visitOrder.append(i)
        linkedWebpage = L[i] - 1
        webpagesVisited = 1
        while linkedWebpage not in visited:
            if linkedWebpage in webpageHops:
                # We've already visited this webpage and calculated the number of webpages that can be visited from it
                webpagesVisited += webpageHops[linkedWebpage]
                break

            # Visit the next webpage
            webpagesVisited += 1
            visited.add(linkedWebpage)
            visitOrder.append(linkedWebpage)
            linkedWebpage = L[linkedWebpage] - 1

        maxVisitableWebpages = max(maxVisitableWebpages, webpagesVisited)

        reached_cycle = False
        # Update the number of webpages that can be visited from each webpage in the visit order
        while visitOrder:
            page = visitOrder.popleft()

            if page == linkedWebpage:
                # We've reached a cycle because the last webpage links back to a webpage we've already visited
                reached_cycle = True

            webpageHops[page] = webpagesVisited

            if not reached_cycle:
                # We haven't reached a cycle yet, so this webpage can visit one less webpage than the webpage that linked to it
                webpagesVisited -= 1

        # Minor optimization to break early if we've visited all webpages
        if len(webpageHops) == N:
            break

    return maxVisitableWebpages